daemon/api.go

```diff
@@ -586,78 +612,151 @@ func deleteOp(c *Command, r *http.Request) Response {
 }
 
 type packageInstruction struct {
+	progress.NullProgress
 	Action   string `json:"action"`
 	LeaveOld bool   `json:"leave_old"`
 	pkg      string
-	prog     progress.Meter
+	chin     chan interface{}
+	chout    chan interface{}
 }
 
-func (inst *packageInstruction) install() interface{} {
-	flags := snappy.DoInstallGC
-	if inst.LeaveOld {
-		flags = 0
+func (inst *packageInstruction) Agreed(intro, licenseFile string) bool {
+	inst.chout <- map[string]interface{}{
+		"license_ack_needed": true,
+		"license_intro":      intro,
+		"license_text":       licenseFile,
 	}
-	_, err := snappy.Install(inst.pkg, flags, inst.prog)
 
+	b, ok := (<-inst.chin).(bool)
+	inst.chout <- nil
+
+	return ok && b
+}
+
+func (inst *packageInstruction) install() interface{} {
+	inst.chin = make(chan interface{})
+	inst.chout = make(chan interface{})
+	lock, err := lockfile.Lock(dirs.SnapLockFile, true)
 	if err != nil {
 		return err
 	}
 
-	// TODO: return the log
-	// also to do: commands update their output dynamically, as it changes.
-	return nil
+	go func() {
+		defer close(inst.chout)
+		defer lock.Unlock()
+
+		flags := snappy.DoInstallGC
+		if inst.LeaveOld {
+			flags = 0
+		}
+		_, err = snappy.Install(inst.pkg, flags, inst)
+
+		inst.chout <- err
+
+		// TODO: return the log
+		// also to do: commands update their output dynamically, as it changes.
+	}()
+
+	return [2]chan interface{}{inst.chin, inst.chout}
 }
 
 func (inst *packageInstruction) update() interface{} {
-	// zomg :-(
-	// TODO: query the store for just this package, instead of this
+	inst.chin = make(chan interface{})
+	inst.chout = make(chan interface{})
 
-	flags := snappy.DoInstallGC
-	if inst.LeaveOld {
-		flags = 0
-	}
-
-	parts, err := snappy.ListUpdates()
+	lock, err := lockfile.Lock(dirs.SnapLockFile, true)
 	if err != nil {
 		return err
 	}
 
-	for _, part := range parts {
-		if snappy.QualifiedName(part) == inst.pkg {
-			if _, err := part.Install(inst.prog, flags); err != nil {
-				return err
+	go func() {
+		// zomg :-(
```

Comment:
There is a merge conflict with the selective-updates branch waiting here, but I guess there is nothing we can do about it